test/functorch/test_eager_transforms.py

```diff
@@ -32,13 +32,13 @@
 from torch._C import _ExcludeDispatchKeyGuard, DispatchKeySet, DispatchKey
 
 import functorch
-from functorch import (
-    grad, vjp, vmap, jacrev, jacfwd, grad_and_value, hessian,
-    jvp, make_functional, make_functional_with_buffers,
-    combine_state_for_ensemble, make_fx
+from torch.func import (
```

Comment:
>  The purpose is that dynamo won't skip _cond.py.

Isn't what you want to move _cond.py from functorch/experimental/_cond.py into torch/_higher_order_ops/* ?

Reply:
yeah, you're right. Let me find out how things goes. 

Given we probably won't need this pr for https://github.com/pytorch/pytorch/pull/105679. Do you think you still want to land it or we should just abandon it?

Reply:
Just found out on master branch, when using the deprecated apis, we have:
```python
import torch
from functorch import grad

@torch.compile(backend="eager")
def test_argnums():
    x = torch.randn([])
    y = torch.randn([])
    gx = grad(torch.mul, argnums=0)(x, y)
    print(gx, y)

    gy = grad(torch.mul, argnums=1)(x, y)
    print(gy, x)

    gx = grad(torch.mul, argnums=(0,))(x, y)
    print(gx, y)

    gx, gy = grad(torch.mul, argnums=(0, 1))(x, y)
    print(gx, y)
    print(gy, x)

test_argnums()
```
This code gives us:
```
(pytorch-3.10) ~/local/pytorch$ python t.py 
[2023-08-02 13:18:58,002] torch._dynamo.convert_frame: [WARNING] WON'T CONVERT test_argnums /home/yidi/local/pytorch/t.py line 4 
[2023-08-02 13:18:58,002] torch._dynamo.convert_frame: [WARNING] due to: 
[2023-08-02 13:18:58,002] torch._dynamo.convert_frame: [WARNING] Traceback (most recent call last):
[2023-08-02 13:18:58,002] torch._dynamo.convert_frame: [WARNING]   File "/home/yidi/local/pytorch/torch/fx/proxy.py", line 274, in create_arg
[2023-08-02 13:18:58,002] torch._dynamo.convert_frame: [WARNING]     raise NotImplementedError(f"argument of type: {type(a)}")
[2023-08-02 13:18:58,002] torch._dynamo.convert_frame: [WARNING] NotImplementedError: argument of type: <class 'builtin_function_or_method'>
[2023-08-02 13:18:58,002] torch._dynamo.convert_frame: [WARNING] 
[2023-08-02 13:18:58,002] torch._dynamo.convert_frame: [WARNING] from user code:
[2023-08-02 13:18:58,002] torch._dynamo.convert_frame: [WARNING]    File "/home/yidi/local/pytorch/t.py", line 8, in test_argnums
[2023-08-02 13:18:58,002] torch._dynamo.convert_frame: [WARNING]     gx = grad(torch.mul, argnums=0)(x, y)
[2023-08-02 13:18:58,002] torch._dynamo.convert_frame: [WARNING] 
[2023-08-02 13:18:58,002] torch._dynamo.convert_frame: [WARNING] Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
[2023-08-02 13:18:58,002] torch._dynamo.convert_frame: [WARNING] 
[2023-08-02 13:18:58,002] torch._dynamo.convert_frame: [WARNING] 
[2023-08-02 13:18:58,002] torch._dynamo.convert_frame: [WARNING] converting frame raised error, suppressing error
/home/yidi/local/pytorch/torch/_functorch/deprecated.py:66: UserWarning: We've integrated functorch into PyTorch. As the final step of the integration, functorch.grad is deprecated as of PyTorch 2.0 and will be deleted in a future version of PyTorch >= 2.3. Please use torch.func.grad instead; see the PyTorch 2.0 release notes and/or the torch.func migration guide for more details https://pytorch.org/docs/master/func.migrating.html
  warn_deprecated('grad')
tensor(1.8946) tensor(1.8946)
/home/yidi/local/pytorch/torch/_functorch/deprecated.py:66: UserWarning: We've integrated functorch into PyTorch. As the final step of the integration, functorch.grad is deprecated as of PyTorch 2.0 and will be deleted in a future version of PyTorch >= 2.3. Please use torch.func.grad instead; see the PyTorch 2.0 release notes and/or the torch.func migration guide for more details https://pytorch.org/docs/master/func.migrating.html
  warn_deprecated('grad')
tensor(1.8946) tensor(-0.4774)
tensor(1.8946) tensor(1.8946)
Traceback (most recent call last):
  File "/home/yidi/local/pytorch/t.py", line 21, in <module>
    test_argnums()
  File "/home/yidi/local/pytorch/torch/_dynamo/eval_frame.py", line 321, in _fn
    return fn(*args, **kwargs)
  File "/home/yidi/local/pytorch/t.py", line 17, in test_argnums
    gx, gy = grad(torch.mul, argnums=(0, 1))(x, y)
  File "/home/yidi/local/pytorch/torch/_tensor.py", line 933, in __iter__
    raise TypeError("iteration over a 0-d tensor")
TypeError: iteration over a 0-d tensor
```
But switching "from functorch import grad" to "from torch.func import grad" produces correct results:
```
(pytorch-3.10) ~/local/pytorch$ python t.py 
tensor(-0.5328) tensor(-0.5328)
tensor(1.2671) tensor(1.2671)
(tensor(-0.5328),) tensor(-0.5328)
tensor(-0.5328) tensor(-0.5328)
tensor(1.2671) tensor(1.2671)
```
, which seems interesting...

Reply:
cc @kshitij12345 -- something weird going on here with Dynamo support for grad? Ideally functorch.grad works the same way as torch.func.grad, because I am not sure how many people have switched over yet

Reply:
Having a look, thanks for the ping.

Reply:
I am unable to repro this with current main: 0dc7f6df9d00428cd175018e2bf9b45a8ec39b9c, 

Output looks fine (deprecations are printed and results match)
```
/home/kshiteej/Pytorch/pytorch_functorch/torch/_functorch/deprecated.py:66: UserWarning: We've integrated functorch into PyTorch. As the final step of the integration, functorch.grad is deprecated as of PyTorch 2.0 and will be deleted in a future version of PyTorch >= 2.3. Please use torch.func.grad instead; see the PyTorch 2.0 release notes and/or the torch.func migration guide for more details https://pytorch.org/docs/master/func.migrating.html
  warn_deprecated('grad')
tensor(0.0892) tensor(0.0892)
/home/kshiteej/Pytorch/pytorch_functorch/torch/_functorch/deprecated.py:66: UserWarning: We've integrated functorch into PyTorch. As the final step of the integration, functorch.grad is deprecated as of PyTorch 2.0 and will be deleted in a future version of PyTorch >= 2.3. Please use torch.func.grad instead; see the PyTorch 2.0 release notes and/or the torch.func migration guide for more details https://pytorch.org/docs/master/func.migrating.html
  warn_deprecated('grad')
tensor(-0.2102) tensor(-0.2102)
/home/kshiteej/Pytorch/pytorch_functorch/torch/_functorch/deprecated.py:66: UserWarning: We've integrated functorch into PyTorch. As the final step of the integration, functorch.grad is deprecated as of PyTorch 2.0 and will be deleted in a future version of PyTorch >= 2.3. Please use torch.func.grad instead; see the PyTorch 2.0 release notes and/or the torch.func migration guide for more details https://pytorch.org/docs/master/func.migrating.html
  warn_deprecated('grad')
(tensor(0.0892),) tensor(0.0892)
/home/kshiteej/Pytorch/pytorch_functorch/torch/_functorch/deprecated.py:66: UserWarning: We've integrated functorch into PyTorch. As the final step of the integration, functorch.grad is deprecated as of PyTorch 2.0 and will be deleted in a future version of PyTorch >= 2.3. Please use torch.func.grad instead; see the PyTorch 2.0 release notes and/or the torch.func migration guide for more details https://pytorch.org/docs/master/func.migrating.html
  warn_deprecated('grad')
tensor(0.0892) tensor(0.0892)
tensor(-0.2102) tensor(-0.2102)
```

Reply:
Not sure what's going on...Will submit an issue if the error persists.

Reply:
I can reproduce this. Latest commit i see is `410bc558e62 - Assert that `args` is of tuple type. (#106352) (15 hours ago) <gmagogsfm>`

Reply:
Weird, even with mentioned commit 410bc558e625efceabbe6d6bcec639826ac41489, it is working for me with deprecation warnings 🤔 
<img width="1208" alt="image" src="https://github.com/pytorch/pytorch/assets/19503980/91c8c061-d235-49a6-9dcb-9dbe9a13d113">

Reply:
I am able to reproduce it, somehow running the script from another location outside of pytorch test folder triggers this.

Reply:
So, what is happening is that `functorch.grad` is treated as `TorchVariable` but we should be tracing through it. Updating `allowed_functions.py` works. Will fix it in a separate PR.

```patch
diff --git a/torch/_dynamo/allowed_functions.py b/torch/_dynamo/allowed_functions.py
index dd36149efdc..95580df5391 100644
--- a/torch/_dynamo/allowed_functions.py
+++ b/torch/_dynamo/allowed_functions.py
@@ -194,7 +194,7 @@ def _allowed_function_ids():
                 if isinstance(obj, torch._ops.HigherOrderOperator):
                     continue
                 # We want to trace through `grad`
-                if obj is torch.func.grad:
+                if obj in (torch.func.grad, torch._functorch.deprecated.grad):
                     continue
 
                 if isinstance(obj, types.ModuleType):
```